### PR TITLE
Reversed the semantics of attaching an interaction 

### PR DESCRIPTION
**Pull request by aicioara**
```typescript
var interaction = new Plotable.Interactions.Pointer();
var component = new Plottable.Component();

component.registerInteraction(interaction); // old
interaction.attachTo(component); //new
```

As part of this commit there are various side effects:
- Having an `onAnchor` and `offAnchor` public methods on the `Component`
- have a symmetric `detachFrom(component)` on `Interaction`
- detaching Interaction requires unsubscribing from `Dispatcher`s, hence callbacks have to be now instance variables
- there was no test coverage on attaching / detaching interactions. Even now the test coverage is pretty low, as it is very hard to test this at an abstract level (abstract interaction has no dispatchers attached), hence only click interaction is tested
- hence created an `interactionTests.ts` file

### CODE REVIEW

File: src/components/component.ts [hunk 1]
```diff
@@ -67,6 +70,36 @@ module Plottable {
...
+    public onAnchor(callback: AnchorCallback) {
+      if (this._isAnchored) {
+        callback(this);
+      }
+      this._onAnchorCallbacks.add(callback);
+      return this;
+    }
+
+    /**
+     * Removes a callback to be called on anchoring the Component to the DOM.
```
Comment (bluong) [inline src/components/component.ts]: Let's match the on/offAnchor docs exactly.

`onAnchor` has the word current.  Let's remove that

File: src/components/component.ts [hunk 2]
```diff
@@ -67,6 +70,36 @@ module Plottable {
...
+     * The callback is identified by reference equality.
+     *
+     * @param {AnchorCallback} callback The callback to be removed
```
Comment (bluong) [inline src/components/component.ts]: Let's have a better parallel between the doc between on/offAnchor

File: src/interactions/interaction.ts [hunk 3]
```diff
@@ -11,8 +11,53 @@ module Plottable {
      */
     protected _componentToListenTo: Component;
 
-    public _anchor(component: Component) {
+    private _anchorCallback = (component: Component) => this._anchor(component);
+
+    private _isAnchored: Boolean;
+
+    /* tslint:disable: */
```
Comment (jtlan) [inline src/interactions/interaction.ts]: Is this still needed? I tried pulling the branch and removing the `disable`, and tslint still passed.

File: src/interactions/interaction.ts [hunk 4]
```diff
@@ -11,8 +11,53 @@ module Plottable {
...
+    // tslint disable does not really work
+    protected _anchor(component: Component) {
+      this._isAnchored = true;
+    }
+
+    protected _unanchor() {
+      this._isAnchored = false;
+    }
+    /* tslint:enable */
```
Comment (bluong) [inline src/interactions/interaction.ts]: What is with the tslint messages here?

File: src/interactions/interaction.ts [hunk 5]
```diff
@@ -11,8 +11,53 @@ module Plottable {
...
+     * attached to a Component, it first detaches itself from the old component.
+     *
+     * @param {Component} component The component to which to attach the interaction.
+     *
+     * @return {Interaction}
+     */
+    public attachTo(component: Component) {
+      if (this._componentToListenTo) {
+        this.detachFrom(this._componentToListenTo);
+      }
```
Comment (bluong) [inline src/interactions/interaction.ts]: Possibly a premature optimization but should this no-op  if
component === this._componentToListenTo

File: src/interactions/interaction.ts [hunk 6]
```diff
@@ -11,8 +11,53 @@ module Plottable {
      */
     protected _componentToListenTo: Component;
```
Comment (bluong) [inline src/interactions/interaction.ts]: Should we rename this to just `_component`?

File: src/interactions/interaction.ts [hunk 7]
```diff
@@ -11,8 +11,53 @@ module Plottable {
...
+
+    /**
+     * Detaches current interaction from the Component. Interaction can be reused.
+     *
+     * @param {Component} component The component to which to attach the interaction.
+     *
+     * @return {Interaction}
+     */
+    public detachFrom(component: Component) {
+      this._unanchor();
```
Comment (bluong) [inline src/interactions/interaction.ts]: Possibly `_unAnchor()`

File: test/interactions/interactionTests.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+      callbackCalled = false;
+      clickInteraction.detachFrom(component);
+
+      TestMethods.triggerFakeMouseEvent("mousedown", component.content(), SVG_WIDTH / 2, SVG_HEIGHT / 2);
+      TestMethods.triggerFakeMouseEvent("mouseup", component.content(), SVG_WIDTH / 2, SVG_HEIGHT / 2);
+      assert.isFalse(callbackCalled, "callback was removed from component and should not be called");
+
+      svg.remove();
+    });
```
Comment (bluong) [inline test/interactions/interactionTests.ts]: Is there a way you can test on the base `Interaction` class?

File: test/interactions/interactionTests.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+      var svg2 = TestMethods.generateSVG(SVG_WIDTH, SVG_HEIGHT);
+      var component1 = new Plottable.Component();
+      var component2 = new Plottable.Component();
+
+      component1.renderTo(svg1);
+      component2.renderTo(svg2);
+
+      var clickInteraction = new Plottable.Interactions.Click();
+
+      var callbackCalled: Boolean;
```
Comment (bluong) [inline test/interactions/interactionTests.ts]: Let's be consistent with the rest of the codebase and use the primitive types.  I am unsure but I thought tslint should ban this

File: test/interactions/interactionTests.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+      TestMethods.triggerFakeMouseEvent("mouseup", component1.content(), SVG_WIDTH / 2, SVG_HEIGHT / 2);
+      assert.isFalse(callbackCalled, "Round 2 callback not called for component 1");
+
+      callbackCalled = false;
+      TestMethods.triggerFakeMouseEvent("mousedown", component2.content(), SVG_WIDTH / 2, SVG_HEIGHT / 2);
+      TestMethods.triggerFakeMouseEvent("mouseup", component2.content(), SVG_WIDTH / 2, SVG_HEIGHT / 2);
+      assert.isTrue(callbackCalled, "Round 2 callback called for component 2");
+
+      // no detach, as this shorthand should also work
+      clickInteraction.attachTo(component1);
```
Comment (bluong) [inline test/interactions/interactionTests.ts]: Instead of making a comment here, can we instead have a test to prove this or it should be decently obvious what you are doing here anyways that the comment is uncecessary

**Comment by bluong** [comment]
A few nitpicky things in the inplementation.  Mostly issues on documentation and test structuring.

File: src/interactions/interaction.ts [hunk 11]
```diff
@@ -11,8 +11,53 @@ module Plottable {
      */
     protected _componentToListenTo: Component;
 
-    public _anchor(component: Component) {
+    private _anchorCallback = (component: Component) => this._anchor(component);
```
Comment (jtlan) [inline src/interactions/interaction.ts]: We'll probably want some sort of mechanism to detach the `Interaction` if the `Component` is `destroy()`ed, so we don't have loose references to the `Component` floating around and preventing garbage collection. However, that's an existing bug, so we can tackle it after the API refactor.

File: src/interactions/interaction.ts (hunk 4 above)
Comment (jtlan) [inline src/interactions/interaction.ts]: Looks like a previous commit had the blocks empty...? I tried pulling the branch in its current state and it worked fine without the tslint stuff.

File: test/interactions/interactionTests.ts (hunk 8 above)
Comment (jtlan) [inline test/interactions/interactionTests.ts]: It looks like he would have to add a dispatcher or something. Would it be enough to spy on `_anchor()` and `_unanchor()`, or `<any>`-cast to check `_componentToListenTo`?

File: test/interactions/interactionTests.ts (hunk 9 above)
Comment (jtlan) [inline test/interactions/interactionTests.ts]: We should probably tslint-ban the non-primitive `Boolean`, `Number`, and `String` types until we can think of a good use for them.

File: src/components/component.ts (hunk 1 above)
Comment (aicioara) [inline src/components/component.ts]: Fixed

File: src/components/component.ts (hunk 2 above)
Comment (aicioara) [inline src/components/component.ts]: I believe fixed, but not sure what you mean

File: src/interactions/interaction.ts (hunk 3 above)
Comment (aicioara) [inline src/interactions/interaction.ts]: @jtlan Intentional leftover, so I don't forget. Basically I need a no-op function (kind of abstract) in the base class. But the tslint does not allow me to do that. So I needed to invent an instance variable which is never ever used. Do you think it is alright? Also tslint disable does not work

File: src/interactions/interaction.ts (hunk 5 above)
Comment (aicioara) [inline src/interactions/interaction.ts]: I believe we're trying to be too smart with that change. The user should not do that in the first place I believe.

File: src/interactions/interaction.ts (hunk 6 above)
Comment (aicioara) [inline src/interactions/interaction.ts]: Not suggestive enough without prior knowledge of how this class works. maybe `_componentAttachedTo` would work (doing this change)

File: src/interactions/interaction.ts (hunk 7 above)
Comment (aicioara) [inline src/interactions/interaction.ts]: `unanchor` is a word in itself (well kind of, unanchored is a word). Camel case when we mean 2 words only

File: test/interactions/interactionTests.ts (hunk 10 above)
Comment (aicioara) [inline test/interactions/interactionTests.ts]: I'd like to avoid writing extra tests to test shorthand for the same thing. Tests are already running for 6 seconds (or more on SauceLabs). I'd also want to make it obvious that the detach was intentional. Not sure if it is without the comment.

File: src/interactions/interaction.ts (hunk 11 above)
Comment (aicioara) [inline src/interactions/interaction.ts]: We probably need an onDestroy callback as well. Will think about this

File: src/components/component.ts [hunk 12]
```diff
@@ -67,6 +70,38 @@ module Plottable {
...
       }
       this._isAnchored = true;
+      this._onAnchorCallbacks.callCallbacks(this);
+      return this;
+    }
+
+    /**
+     * Adds a callback to be called on anchoring the Component to the DOM.
+     * If the component is already anchored, the callback is called immediately.
+     * To later deregister, use offAnchor(callback).
```
Comment (jtlan) [inline src/components/component.ts]: I don't think it's worth cross-referencing the other method, since the point of the `onEvent()`/`offEvent()` naming scheme is to make things sufficiently intuitive for users.

File: test/interactions/interactionTests.ts [hunk 13]
```diff
@@ -0,0 +1,110 @@
+///<reference path="../testReference.ts" />
+
+var assert = chai.assert;
+
+describe("Interactions", () => {
+  describe("Interactions", () => {
```
Comment (jtlan) [inline test/interactions/interactionTests.ts]: Why the double nesting?

File: test/interactions/interactionTests.ts [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+      clickInteraction.detachFrom(component);
+
+      TestMethods.triggerFakeMouseEvent("mousedown", component.content(), SVG_WIDTH / 2, SVG_HEIGHT / 2);
+      TestMethods.triggerFakeMouseEvent("mouseup", component.content(), SVG_WIDTH / 2, SVG_HEIGHT / 2);
+      assert.isFalse(callbackCalled, "callback was removed from component and should not be called");
+
+      svg.remove();
+    });
+
+    it("interactions are reusable", () => {
```
Comment (jtlan) [inline test/interactions/interactionTests.ts]: I think it would be better to split this into three tests:
- `attachTo()`
- `detachFrom()`
- `attachTo() disconnects from previous Component`

File: src/components/component.ts (hunk 12 above)
Comment (aicioara) [inline src/components/component.ts]: Then I did not understand @bluong 's comment

File: src/interactions/interaction.ts [hunk 15]
```diff
@@ -9,10 +9,52 @@ module Plottable {
      * "foreground" and "background" elements where it can draw things,
      * e.g. crosshairs.
      */
-    protected _componentToListenTo: Component;
+    protected _componentAttachedTo: Component;
```
Comment (jtlan) [inline src/interactions/interaction.ts]: As @bluong mentioned, "`_component`" is a sufficient name.

**Comment by jtlan** [comment]
[-1] Test semantics.

File: test/interactions/interactionTests.ts (hunk 13 above)
Comment (aicioara) [inline test/interactions/interactionTests.ts]: I actually meant `Interaction` for the second one. Do you mind double nesting with this renaming?

File: src/interactions/interaction.ts (hunk 15 above)
Comment (aicioara) [inline src/interactions/interaction.ts]: The commenting system here is pretty hard to follow. Pasting my argument here as well

Not suggestive enough without prior knowledge of how this class works. maybe `_componentAttachedTo` would work (doing this change)

File: test/interactions/interactionTests.ts (hunk 14 above)
Comment (aicioara) [inline test/interactions/interactionTests.ts]: I disagree. I believe it's extra boilerplate and extra time consumption for testing (due to generating SVGs, attaching them to the DOM). Also the `detachFrom` test includes the `attachTo` and `attachTo() disconnects from previous Component` is copy+pasting half of the behavior of item 2

Comment (bluong) [inline test/interactions/interactionTests.ts]: I think I agree with Justin in this case.  At our current stage, we should I think be not concerned with the performance of our tests but rather the correctness + maintainability of them.  We shouldn't group a bunch of tests into 1 big test as that detracts from comprehension and independence of the tests

Comment (aicioara) [inline test/interactions/interactionTests.ts]: Performance of the tests should always be a concern. The entire dev process gets slower with slower tests. What is even worse is that we never refactor tests. We always refactor code.

Comment (bluong) [inline test/interactions/interactionTests.ts]: Yes it is indeed a concern, but I feel that there is a higher priority on us having confidence that our library works in the first place and the place that builds that confidence is the tests

**Comment by aicioara** [comment]
The two main points addressed were
`_componentAttachedTo` -> `_component`
- pro: axis does the same for the scale (it does not call it scale attached to)
- cons: _componentAttachedTo is more descriptive. We should allow naming to be self explainatory. We should not read the code (or comments) to figure out what an instance variable is / should do

Separate the attach/detach test into multiple tests
- pro: modularity of the test. concepts of unit tests. other purity stuff...
- pro: I can test attach more thoroughly (throw junk at it, see what it does)
- cons: boilerplate
- cons: time consuming for the tests. Performance of the tests should always be a concern. The entire dev process gets slower with slower tests. What is even worse is that we never refactor tests. We always refactor code.
- cons: we cannot have `detach` without `attach` so it makes no sense to separate the two methods in testing. 
- cons: the fix for this is 1 extra test, which is exactly a subset of the tests I wrote already (well, almost). The existing tests will not shrink anyway.
- cons: there is no modularity anyway. If the attach fails, you can see that in the assert now. If it fails in the split, it will just fail multiple tests, so it is hard to see which one is the issue anyway, because tests are not hierarchical. It's not like the tests know which axioms they are dependent on and where they have been asserted

File: test/interactions/interactionTests.ts (hunk 14 above)
Comment (jtlan) [inline test/interactions/interactionTests.ts]: @aicioara wrote:
>What is even worse is that we never refactor tests. We always refactor code.
We probably *should* refactor tests, and I think that should be something we do immediately after finishing the API refactor.

File: test/interactions/interactionTests.ts [hunk 16]
```diff
@@ -0,0 +1,110 @@
+///<reference path="../testReference.ts" />
+
+var assert = chai.assert;
+
+describe("Interactions", () => {
+  describe("Interaction", () => {
```
Comment (jtlan) [inline test/interactions/interactionTests.ts]: I was going to say that we didn't need this inner `describe()` at all, since anything on the top level (directly under `describe("Interactions")`) could probably be safely assumed to test the root class.

Comment (aicioara) [inline test/interactions/interactionTests.ts]: It's a convenience. If you want to run just the tests in this file, you can click on the `Interaction` in the tests interface. (I very often want to do that). If you put it at the root, there is no way to do it

Comment (jtlan) [inline test/interactions/interactionTests.ts]: Okay, that makes a lot of sense. It might be good to come up with some way to identify the base class in tests, but I think we can save that decision for when we revisit tests.

File: test/interactions/interactionTests.ts (hunk 14 above)
Comment (aicioara) [inline test/interactions/interactionTests.ts]: if we _should_ revisit the tests, then why not bring up the discussion again then and leave the tests like this for now.

**Comment by bluong** [comment]
Decently important preface - Our testing infrastructure for essentially the last year (or more) has essentially been using a pure additive structure.  (Add a feature?  Add a test.), but it hasn't necessarily been made concise in almost any way.

One main instance is that there are many tests that are trying to prove a fact that other tests have already proven in a much shorter, performant fashion.  (scatterPlots work on quadratic accessors! or something to that effect)

Thus, we definitely need to improve that piece in our codebase for the reasons you've described but it's also important on our end that we cleanup the tests so we have a much better understanding of what our library actually does

In accordance with the above preface, I still think the test should be broken up, mainly because I see the test name and I don't think it's very unitized. "interactions are reusable"

The test names should point near directly at the fact they are trying to prove and the test code should follow

File: test/interactions/interactionTests.ts (hunk 16 above)
Comment (bluong) [inline test/interactions/interactionTests.ts]: The tests interface is extremely thin if you look at the url.  It literally does a grep for the string you pressed and runs those according tests

File: test/interactions/interactionTests.ts (hunk 14 above)
Comment (dbt) [inline test/interactions/interactionTests.ts]: If you know something is the right thing to do, why wait?

the performance of the tests is a relatively small concern, and we can address performance issues if they extend the suite running time by a significant fraction.

If you're worried about duplication of code between the attach and detach tests, you can pull it out into a local function that you invoke both times.

Comment (aicioara) [inline test/interactions/interactionTests.ts]: @dbt because I don't believe it is the right thing to do. However, I was a minority so made the change.

File: plottable.d.ts [hunk 17]
```diff
@@ -1502,6 +1503,24 @@ declare module Plottable {
...
+         * If the component is already anchored, the callback is called immediately.
+         *
+         * @param {AnchorCallback} callback The callback to be called on Anchor
+         *
+         * @return {Component}
+         */
+        onAnchor(callback: AnchorCallback): Component;
+        /**
+         * Removes a callback to be called on anchoring the Component to the DOM.
+         * The callback is identified by reference equality.
```
Comment (bluong) [inline plottable.d.ts]: Do you think the above line is necessary?

File: plottable.d.ts [hunk 18]
```diff
@@ -1502,6 +1503,24 @@ declare module Plottable {
          */
         anchor(selection: D3.Selection): Component;
         /**
+         * Adds a callback to be called on anchoring the Component to the DOM.
+         * If the component is already anchored, the callback is called immediately.
+         *
+         * @param {AnchorCallback} callback The callback to be called on Anchor
```
Comment (bluong) [inline plottable.d.ts]: lowercase anchor

File: plottable.d.ts [hunk 19]
```diff
@@ -1502,6 +1503,24 @@ declare module Plottable {
...
+         *
+         * @return {Component}
+         */
+        onAnchor(callback: AnchorCallback): Component;
+        /**
+         * Removes a callback to be called on anchoring the Component to the DOM.
+         * The callback is identified by reference equality.
+         *
+         * @param {AnchorCallback} callback The callback to be removed
```
Comment (bluong) [inline plottable.d.ts]: I'm trying to find a better parallel between the `onAnchor` and `offAnchor` docs and it mostly looks good.

For this param part, would it be too vague to say
The callback to be added / removed ?

I guess another way to word it is that the parallel doc for `offAnchor` are a little odd, `The callback to be called off Anchor`

File: plottable.d.ts [hunk 20]
```diff
@@ -3680,15 +3692,26 @@ declare module Plottable {
...
+        protected _componentAttachedTo: Component;
+        protected _anchor(component: Component): void;
+        protected _unanchor(): void;
         /**
-         * It maintains a 'hitBox' which is where all event listeners are
-         * attached. Due to cross- browser weirdness, the hitbox needs to be an
-         * opaque but invisible rectangle.  TODO: We should give the interaction
-         * "foreground" and "background" elements where it can draw things,
-         * e.g. crosshairs.
+         * Attaches current interaction to a Component. If the interaction was already
```
Comment (bluong) [inline plottable.d.ts]: Let's put the second sentence on the second line

File: plottable.d.ts [hunk 21]
```diff
@@ -3680,15 +3692,26 @@ declare module Plottable {
...
+         * attached to a Component, it first detaches itself from the old component.
```
Comment (bluong) [inline plottable.d.ts]: Let's put the second sentence on the second line

File: plottable.d.ts (hunk 17 above)
Comment (aicioara) [inline plottable.d.ts]: It depends.

File: plottable.d.ts (hunk 20 above)
Comment (bluong) [inline plottable.d.ts]: I don't this the word `current` here is nice in the doc... I would remove the word completely or replace it with `this`

File: plottable.d.ts [hunk 22]
```diff
@@ -3680,15 +3692,26 @@ declare module Plottable {
...
+         *
+         * @param {Component} component The component to which to attach the interaction.
+         *
+         * @return {Interaction}
+         */
+        attachTo(component: Component): Interaction;
+        /**
+         * Detaches current interaction from the Component. Interaction can be reused.
```
Comment (bluong) [inline plottable.d.ts]: "This interaction can be reused." and also consider placing this on a second line

File: plottable.d.ts [hunk 23]
```diff
@@ -3680,15 +3692,26 @@ declare module Plottable {
...
+         *
+         * @param {Component} component The component to which to attach the interaction.
```
Comment (bluong) [inline plottable.d.ts]: "The component from which to detach the interaction."

File: plottable.d.ts (hunk 19 above)
Comment (aicioara) [inline plottable.d.ts]: I liked the `callback to be added / removed` idea. Changing to that to have better symmetry

**Comment by bluong** [comment]
Docs docs docs
